### ISSUE TITLE
WinSock with OpenSSL error

### ISSUE THREAD

**Issue by AlexWhiteS** [openssl/openssl#20822]
The error gives me the following - error:0a000410:ssl routines::sslv3 alert handshake failure
I did the following steps before creating this thread.
1. Tried different versions of OPENSSL(1, 3)
2. Generated new certificates with openssl
3. Checked the communication protocol - it matches the client and servers
4. Checked the encryption method - it also matches the client and server

An interesting note is that before the client closes with an error, the server manages to accept this error and display it in its console

code client:
```` C
#include <winsock2.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <iostream>
#include <string>
#define SERVER_ADDRESS "127.0.0.1"
#define SERVER_PORT 8080

int main()
{
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        perror("wsadata");
        return 1;
    }

    SSL_library_init();
    OpenSSL_add_all_algorithms();
    SSL_load_error_strings();

    SSL_CTX* ctx = SSL_CTX_new(TLS_client_method());
    SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_NO_COMPRESSION);
    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
    SSL_CTX_load_verify_locations(ctx, "server.crt", NULL);

    SOCKET sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (sock == INVALID_SOCKET)
    {
        perror("socket");
        return -1;
    }

    struct sockaddr_in addr = { 0 };
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = inet_addr(SERVER_ADDRESS);
    addr.sin_port = htons(SERVER_PORT);

    if (connect(sock, (struct sockaddr*)&addr, sizeof(addr)) == SOCKET_ERROR)
    {
        perror("connect");
        return -1;
    }

    SSL* ssl = SSL_new(ctx);
    SSL_set_fd(ssl, sock);


    if (SSL_connect(ssl) != 1) //error here. SSL_connect return -1 and perror return "SSL_connect : No error"
    {
        unsigned long err = ERR_get_error();
        char buf[256];
        ERR_error_string_n(err, buf, sizeof(buf));
        printf("SSL_connect failed with error: %s\n", buf);
        perror("SSL_connect");
        SSL_free(ssl);
        closesocket(sock);
        
    }

    const char* message = "Hello, server!";
    int len = SSL_write(ssl, message, strlen(message));
    if (len <= 0)
    {
        perror("SSL_write");
        SSL_shutdown(ssl);
        SSL_free(ssl);
        closesocket(sock);
        return -1;
    }

    char buf[1024];
    len = SSL_read(ssl, buf, sizeof(buf) - 1);
    if (len <= 0)
    {
        perror("SSL_read");
        SSL_shutdown(ssl);
        SSL_free(ssl);
        closesocket(sock);
        return -1;
    }

    buf[len] = '\0';
    printf("Server : %s\n", buf);

    SSL_shutdown(ssl);
    SSL_free(ssl);
    closesocket(sock);

    SSL_CTX_free(ctx);

    return 0;
}
````

code server - the server is working well
```` C
#define _CRT_SECURE_NO_WARNINGS
#include <winsock2.h>
#include <openssl/ssl.h>
#include <openssl/err.h>

#define SERVER_PORT 8080

int main()
{
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        perror("wsadata");
        return 1;
    }

    SSL_library_init();
    OpenSSL_add_all_algorithms();
    SSL_load_error_strings();

    SSL_CTX* ctx = SSL_CTX_new(TLS_server_method());
    SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3 | SSL_OP_NO_COMPRESSION);
    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, NULL);
    SSL_CTX_use_PrivateKey_file(ctx, "server.key", NULL);
    SSL_CTX_use_certificate_file(ctx, "server.crt", NULL);

    SOCKET sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (sock == INVALID_SOCKET)
    {
        perror("socket");
        return -1;
    }

    struct sockaddr_in addr = { 0 };
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_ANY);
    addr.sin_port = htons(SERVER_PORT);
    if (bind(sock, (struct sockaddr*)&addr, sizeof(addr)) == SOCKET_ERROR)
    {
        perror("bind");
        return -1;
    }

    if (listen(sock, SOMAXCONN) == SOCKET_ERROR)
    {
        perror("listen");
        return -1;
    }

    while (true)
    {
        struct sockaddr_in client_addr = { 0 };
        int addr_len = sizeof(client_addr);
        SOCKET client_sock = accept(sock, (struct sockaddr*)&client_addr, &addr_len);
        if (client_sock == INVALID_SOCKET)
        {
            perror("accept");
            continue;
        }

        SSL* ssl = SSL_new(ctx);
        SSL_set_fd(ssl, client_sock);

        if (SSL_accept(ssl) != 1)
        {
            perror("SSL_accept");
            SSL_free(ssl);
            closesocket(client_sock);
            continue;
        }

        char buf[1024];
        int len = SSL_read(ssl, buf, sizeof(buf) - 1);
        if (len <= 0)
        {
            perror("SSL_read");
            SSL_shutdown(ssl);
            SSL_free(ssl);
            closesocket(client_sock);
            continue;
        }

        buf[len] = '\0';
        printf("Client says: %s\n", buf);

        const char* message = "Hello, client!";
        len = SSL_write(ssl, message, strlen(message));
        if (len <= 0)
        {
            perror("SSL_write");
        }

        SSL_shutdown(ssl);
        SSL_free(ssl);
        closesocket(client_sock);
    }

    SSL_CTX_free(ctx);

    return 0;
}
````

**Comment by mattcaswell** [comment]
Please show us the output you get on the server side when you see this error on the client side.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
